diff --git a/FaceAlignment/src/cfan.cpp b/FaceAlignment/src/cfan.cpp
index 47f09ae..3d65814 100644
--- a/FaceAlignment/src/cfan.cpp
+++ b/FaceAlignment/src/cfan.cpp
@@ -205,7 +205,7 @@ void CCFAN::FacialPointLocate(const unsigned char *gray_im, int im_width, int im
   {
     for (int j = 0; j < pts_num_; j++)
     {
-      if (isnan(fea[j * 128 + i]))
+      if (std::isnan(fea[j * 128 + i]))
       {
         re_fea[i*pts_num_ + j] = 0;
       }
@@ -281,7 +281,7 @@ void CCFAN::FacialPointLocate(const unsigned char *gray_im, int im_width, int im
   {
     for (int j = 0; j < pts_num_; j++)
     {
-      if (isnan(fea[j * 128 + i]))
+      if (std::isnan(fea[j * 128 + i]))
       {
         re_fea[i*pts_num_ + j] = 0;
       }
